amads/core/vectors_sets.py

```diff
@@ -0,0 +1,294 @@
+    (0, 1, 1, 1, 0, 0, 0)
+
+    >>> set_to_vector(test_set, min_index=1, max_index=6)
+    (1, 1, 1, 0, 0, 0)
+
+    >>> set_to_vector(test_set, max_index=None)
+    (0, 1, 1, 1)
+
+    """
+    if not max_index:
+        max_index = max(input_set)
+    counts = [0] * (max_index + 1)
+    for item in input_set:
+        counts[item] += 1
+
+    if min_index > 0:
+        counts = counts[min_index:]
+
+    return tuple(counts)
+
+
+def vector_to_set(vector: list[int] | tuple[int]) -> tuple:
+    """
+    Converts any "vector" (count of integers organised by index)
+    to a corresponding "set" (unordered list of integers).
+    See the paper for full definitions.
+
+    Parameters
+    ----------
+    vector: The input vector.
+
+    Returns
+    -------
+    tuple: The corresponding set.
+
+    Examples
+    --------
+    >>> test_vector = (0, 3, 2, 1, 0, 0, 0)
+    >>> resulting_set = vector_to_set(test_vector)
+    >>> resulting_set
+    (1, 1, 1, 2, 2, 3)
+
+    >>> roundtrip = set_to_vector(resulting_set, max_index=6)
+    >>> roundtrip
+    (0, 3, 2, 1, 0, 0, 0)
+
+    >>> roundtrip == test_vector
+    True
+
+    """
+    return tuple(i for i in range(len(vector)) for _ in range(vector[i]))
+
+
+# Arithmetic operations: Addition/subtraction, multiplication,division
+
+
+def apply_constant(
+    set_or_vector: list[int] | tuple[int], constant: float, modulo: int | None = None
+) -> list | tuple:
+    """
+    Apply a constant value to a set or vector.
+
+    Parameters
+    ----------
+    set_or_vector: Any list or tuple representing a set or vector.
+    constant: The constant may be an int or a float and positive or negative.
+    modulo: Int or `None`. If int, return the values modulo this value (e.g., 12 for pitch class sets)
+
+    Returns
+    -------
+    list | tuple: The set or vector with constant applied.
+
+    Examples
+    --------
+
+    >>> start = [0, 1, 2, 3, 11]
+    >>> more = apply_constant(start, 1)
+    >>> more
+    [1, 2, 3, 4, 12]
+
+    >>> less = apply_constant(more, -1)
+    >>> less
+    [0, 1, 2, 3, 11]
+
+    >>> start == less
+    True
+
+    >>> more = apply_constant(start, 1, modulo=12)
+    >>> more
+    [1, 2, 3, 4, 0]
+    """
+    result = [x + constant for x in set_or_vector]
+    if modulo is not None:
+        result = [x % modulo for x in result]
+    return result
+
+
+def scalar_multiply(input_list: list, scale_factor: int = 2) -> list:
+    """
+    Multiply all values of a list.
+
+    Parameters
+    ----------
+    input_list: Any list.
+    scale_factor: The "scale factor" aka "multiplicative operand". Multiply all terms by this amount. Defaults to 2.
+
+    >>> scalar_multiply([0, 1, 2])
+    [0, 2, 4]
+
+    """
+    return [
+        x * scale_factor for x in input_list
+    ]  # TODO np would be better in cases like this
+
+
+# Transformations
+
+
+def rotate(vector: list[int] | tuple[int], steps: int | None = None) -> list:
+    """
+    Rotate a list by N steps.
```

Comment:
Maybe 'rotate a vector'? I think it is good to be consistent about what we are calling our input object